hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java

```diff
@@ -730,6 +739,83 @@ public QueryImplementor createQuery(String queryString) {
 		}
 	}
 
+	private CriteriaCompiler criteriaCompiler;
+
+	@SuppressWarnings("WeakerAccess")
+	protected CriteriaCompiler criteriaCompiler() {
+		if ( criteriaCompiler == null ) {
```

Comment:
This is never called in a concurrent way?

Reply:
I'm 100% sure it's okay for `Session`, and for `StatelessSession` I don't think this class is thread-safe anyway: we use this pattern in other places, such as `org.hibernate.internal.AbstractSharedSessionContract#accessTransaction` or `org.hibernate.internal.AbstractSharedSessionContract#getSessionIdentifier`.

Even if this was called in a concurrent way, the worst that could happen is that we create the `CriteriaCompiler` twice. As far as I can tell, this would work just fine, since `CriteriaCompiler` is stateless.

Reply:
As discussed, `org.hibernate.internal.AbstractSharedSessionContract#accessTransaction` suffers from the same issue and is exposed (indirectly) through the `StatelessSession` interface, so it seems `StatelessSession` is not thread-safe anyway. Plus there's the matter of the connection, which probably cannot be used from multiple threads simultaneously.